include/rsl/parameter_validators.hpp

```diff
@@ -0,0 +1,302 @@
+                                std::string const& predicate_description, Fn const& predicate)
+    -> tl::expected<void, std::string> {
+    switch (parameter.get_type()) {
+        case rclcpp::ParameterType::PARAMETER_STRING:
+            if (auto value = parameter.get_value<std::string>(); !predicate(value.size(), size))
+                return tl::make_unexpected(
+                    fmt::format("Invalid length '{}' for parameter '{}'. Required {}: {}",
+                                value.size(), parameter.get_name(), predicate_description, size));
+            break;
+        default:
+            if (auto value = parameter.get_value<std::vector<T>>(); !predicate(value.size(), size))
+                return tl::make_unexpected(
+                    fmt::format("Invalid length '{}' for parameter '{}'. Required {}: {}",
+                                value.size(), parameter.get_name(), predicate_description, size));
+    }
+    return {};
+}
+
+template <typename T, typename Fn>
+[[nodiscard]] auto compare(rclcpp::Parameter const& parameter, T const& value,
+                           std::string const& predicate_description, Fn const& predicate)
+    -> tl::expected<void, std::string> {
+    if (auto const param_value = parameter.get_value<T>(); !predicate(param_value, value))
+        return tl::make_unexpected(
+            fmt::format("Invalid value '{}' for parameter '{}'. Required {}: {}", param_value,
+                        parameter.get_name(), predicate_description, value));
+    return {};
+}
+}  // namespace detail
+/**
+ * @endcond
+ */
+
+/** @file Validation functions for rclcpp::Parameter */
+
+/**
+ * @brief Is every element of rclcpp::Parameter unique?
+ * @pre rclcpp::Parameter must be an array type
+ * @tparam T the interior type of array; Example: for parameter type double_array, T = double
+ * @return help string if the parameter is invalid, otherwise void
+ */
+template <typename T>
+[[nodiscard]] auto unique(rclcpp::Parameter const& parameter) -> tl::expected<void, std::string> {
+    if (is_unique(parameter.get_value<std::vector<T>>())) return {};
+    return tl::make_unexpected(
+        fmt::format("Parameter '{}' must only contain unique values", parameter.get_name()));
+}
+
+/**
+ * @brief Are the values in parameter a subset of the valid values?
+ * @pre rclcpp::Parameter must be an array type
+ * @tparam T the interior type of array; Example: for parameter type double_array, T = double
+ * @return help string if the parameter is invalid, otherwise void
+ */
+template <typename T>
+[[nodiscard]] auto subset_of(rclcpp::Parameter const& parameter, std::vector<T> const& valid_values)
+    -> tl::expected<void, std::string> {
+    auto const& values = parameter.get_value<std::vector<T>>();
+    for (auto const& value : values)
+        if (!contains(valid_values, value))
+            return tl::make_unexpected(
+                fmt::format("Invalid entry '{}' for parameter '{}'. Not in set: {}", value,
```

Comment:
```suggestion
                fmt::format("Entry '{}' in parameter '{}' not in set {{}}", value,
```

Reply:
I'd prefer a more imperative tense like this. I also would like to wrap sets of elements in curly braces to help distinguish them from the text around them.